src/posts/2021-01-20-ebpf-updates/index.md

```diff
@@ -0,0 +1,315 @@
+
+* Support retrieval of more `SOL_SOCKET` level options from `sock_addr` eBPF
+  programs, to fill the gap between the list of options that `bpf_setsockopt()`
+  can set and those that `bpf_getsockopt()` could retrieve. The concerned
+  options are:
+  * `SO_MARK`
+  * `SO_PRIORITY`
+  * `SO_BINDTOIFINDEX` (also new for `bpf_setsockopt()`)
+
+  (Daniel Borkmann,
+  [link](https://lore.kernel.org/bpf/cba44439b801e5ddc1170e5be787f4dc93a2d7f9.1610406333.git.daniel@iogearbox.net/t/#u))
+
+* Improve out-of-tree cross-building for eBPF selftests. Although this adds no
+  new feature, it feels interesting to report because it should enable wider
+  automated testing on ARM architectures. Selftests are, of course, an
+  essential part of the eBPF ecosystem.
+  (Jean-Philippe Brucker,
+  [link](https://lore.kernel.org/bpf/20210113163319.1516382-1-jean-philippe@linaro.org/t/#u))
+
+## Did You Know?: eBPF Virtual Filesystem
+
+eBPF objects, such as a program or a map, reside in kernel memory until they
+are no longer needed. Internally, the kernel uses reference counters to keep
+track of the number of “handles” pointing to such objects. When the number of
+references comes down to zero, the program or the map is destroyed. The
+references to a program would typically be a hook where the user has attached
+the program (such as a tc filter or a kernel probe), or file descriptors to the
+program that were returned from the kernel when loading the program with the
+`bpf()` system call. Similarly, references to an eBPF map can be held by eBPF
+programs using the map or by a user program that retrieved a file descriptor.
+
+This means that if a process loaded an eBPF program but has not attached it,
+the program will be destroyed when the process exits and its file descriptors
+are closed. There are ways to share file descriptors between processes, but to
+make it easier to reference eBPF objects between user applications, or simply
+to make them persistent at a time when they have no reference in the kernel
+(such as a detached program or an unused map), another mechanism has been
+created: the eBPF virtual filesystem.
+
+The eBPF virtual (or _pseudo_) filesystem, often called bpffs, is traditionally
+mounted at `/sys/fs/bpf`, but any alternative mount point can work. It is
+possible to pin objects to this virtual filesystem, which is rendered as file
+paths. Calling the `bpf()` system call with its `BPF_OBJ_PIN` subcommand pins
+an object. Then, using the `BPF_OBJ_GET` subcommand on a bpffs path retrieves a
+file descriptor to this pinned object. Removing a pinned path simply involves a
+call to `unlink()`, just like for regular paths. Pinned paths (and the eBPF
+objects they reference) are not persistent after reboot.
+
+Note that the use of periods (`.`) in pinned paths is restricted. The glyph has
+long been unused, but a recent feature introduced it to mark paths to specific
+eBPF iterators than the system can preload, `maps.debug` and `progs.debug` (but
```

Comment:
```suggestion
eBPF iterators which the system can preload, such as `maps.debug` and `progs.debug` (but
```
Although, it may be best to just remove that:
```suggestion
eBPF iterators, such as `maps.debug` and `progs.debug` (but
```

Reply:
I don't think there are cases other than `maps.debug` and `progs.debug` at the moment, and I also don't think you can have a period in path names for regular iterators. I don't know if you can load maps.debug/progs.debug after boot, though.

I'll definitely fix the `than`, I probably thought `that`.

Reply:
With `that`, you're saying that **only** iterators the system can preload are allowed to contain a dot in their corresponding filename. Is that what you're trying to say?

Reply:
Sorry, I missed that answer - Yes, I think that only maps.debug and progs.debug can have periods, and both can be preloaded (I don't know if they can be used without preloading). I don't believe there is any other iterator, and in particular any one that could not be preloaded, that can take a period in its pinned path at this time.